diff --exclude autom4te.cache --exclude '*.o' --exclude 'config.*' --exclude Makefile --exclude 'Makefile-*' --exclude '*.orig' --exclude version.c --exclude tcpdump.1 --exclude tcpdump --exclude print-ip45.c --exclude ip45.h -rupN tcpdump/Makefile.in tcpdump-ip45/Makefile.in
--- tcpdump/Makefile.in	2009-09-21 19:16:14.000000000 +0200
+++ tcpdump-ip45/Makefile.in	2013-03-14 21:26:17.000000000 +0100
@@ -75,6 +75,7 @@ CSRC =	addrtoname.c af.c checksum.c cpac
 	print-dccp.c print-decnet.c \
 	print-domain.c print-dtp.c print-dvmrp.c print-enc.c print-egp.c \
 	print-eap.c print-eigrp.c\
+	print-ip45.c \
 	print-esp.c print-ether.c print-fddi.c print-fr.c \
 	print-gre.c print-hsrp.c print-icmp.c print-igmp.c \
 	print-igrp.c print-ip.c print-ipcomp.c print-ipfc.c \
diff --exclude autom4te.cache --exclude '*.o' --exclude 'config.*' --exclude Makefile --exclude 'Makefile-*' --exclude '*.orig' --exclude version.c --exclude tcpdump.1 --exclude tcpdump --exclude print-ip45.c --exclude ip45.h -rupN tcpdump/print-ip.c tcpdump-ip45/print-ip.c
--- tcpdump/print-ip.c	2009-09-21 19:16:14.000000000 +0200
+++ tcpdump-ip45/print-ip.c	2013-07-03 20:25:56.026340151 +0200
@@ -39,6 +39,7 @@ static const char rcsid[] _U_ =
 #include "extract.h"			/* must come after interface.h */
 
 #include "ip.h"
+#include "ip45.h"
 #include "ipproto.h"
 
 struct tok ip_option_values[] = {
@@ -563,6 +564,8 @@ ip_print(netdissect_options *ndo,
 	u_int16_t sum, ip_sum;
 	struct protoent *proto;
 
+	const struct ip45hdr *ip45 = (const struct ip45hdr *)bp;
+
 	ipds->ip = (const struct ip *)bp;
 	if (IP_V(ipds->ip) != 4) { /* print version if != 4 */
 	    printf("IP%u ", IP_V(ipds->ip));
@@ -677,6 +680,12 @@ ip_print(netdissect_options *ndo,
 		ipds->cp = (const u_char *)ipds->ip + hlen;
 		ipds->nh = ipds->ip->ip_p;
 
+		if (is_ip45_pkt(ip45)) {
+			ip45_print(ip45, length);
+			ipds->nh = ip45->nexthdr;
+			ipds->cp = (const u_char *)ip45 + sizeof(struct ip45hdr);
+		}
+
 		if (ipds->nh != IPPROTO_TCP && ipds->nh != IPPROTO_UDP &&
 		    ipds->nh != IPPROTO_SCTP && ipds->nh != IPPROTO_DCCP) {
 			(void)printf("%s > %s: ",
